documentation/puppet-api/v4/compile.markdown

```diff
@@ -0,0 +1,93 @@
+---
+layout: default
+title: "Puppet Server: Puppet API: Compile"
+canonical: "/puppetserver/latest/puppet-api/v4/compile.html"
+---
+
+The compile endpoint allows compilation of arbitrary pcore serialized Puppet
+Code ASTs (aka parsed, but unevaluated Puppet Code snippets). In addition to
+the code AST, the caller must provide the certname, environment, facts,
+trusted facts, and variables for compilation (the context in which this AST
+will be evaluated). Optionally, the caller may provide job_id, transaction_id,
+and configure returned logging. The request body must be JSON formatted and
+the caller must accept a JSON response. The server must not have rich_data
+disabled.
+
+
+## `POST /puppet/v4/compile`
+
+(Introduced in Puppet Server 6.5.0)
+
+The request body must look like:
+```
+{
+  "code_ast": "<pcore serialized ast",
+  "certname": "<node name>",
+  "environment": "<environment name>",
+  "facts": { "values": { "<fact name>": <fact value>, ... } },
+  "trusted_facts": { "values": { "<fact name>": <fact value>, ... } },
+  "variables": { "values": { "<variable name>": <variable value>, ... } },
+
+  # The rest are optional:
+  "transaction_uuid": "<uuid string>",
+  "job_id": "<id string>",
+  "options": { "capture_logs": <true/false>,
+               "log_level": <one of debug/info/warn>}
+}
+```
+
+#### `code_ast` (required)
+A parsed string of json encoded
+[PCore Generic Data](https://github.com/puppetlabs/puppet-specifications/blob/a87c9967ebd7071e0a548920f3262ac24cc3f4cd/language/data-types/pcore-generic-data.md)
+objects representing a Puppet Code AST, as provided in code giving a Puppet AST or Program to
+[Puppet::Pops::Serialization::ToDataConverter](https://github.com/puppetlabs/puppet/blob/68584eaf62361a47da8d825525a3423af1a388c4/lib/puppet/pops/serialization/to_data_converter.rb)
+which represents an intermediate step when compiling Puppet Code to a catalog
+when `rich_data = true` (the default in Puppet 6). The returned catalog is
+equivalent to a catalog returned with the MIME type
+"application/vnd.puppet.rich+json" normally.
```

Comment:
This is definitely a little hard to parse. Maybe it could be simplified a bit. Something like:
```
A parsed string of json encoded PCore Generic Data objects representing a Puppet Code AST. The When `rich_data` is enabled (the default in Puppet 6), the AST represents an intermediate step when compiling Puppet Code to a catalog. The returned catalog is equivalent to a catalog returned with the MIME type "application/vnd.puppet.rich+json" normally.
```

Reply:
I like that better, updated!